### ISSUE TITLE
Extract knn-graph for resulting embeddings

### ISSUE THREAD

**Issue by buhrmann** [lmcinnes/umap#197]
Hi, just a quick question: the nearest neighbours for each output embedding are not already calculated internally and accessible somehow after a fit(), right? (As opposed to the knn in input space). If I wanted to calculate these myself, would the following be the right way?

```python
embeddings = umap.UMAP(n_neighbors=100, min_dist=0.5).fit_transform(X)
knn = umap.umap_.nearest_neighbors(embeddings, n_neighbors=N, 
                                   metric='euclidean', metric_kwds={},
                                   angular=True, random_state=np.random.RandomState(42))
```

**Comment by lmcinnes** [comment]
That would certainly work, however since you are reducing to a very low dimension (2 in this case, since that's the default) you can use sklearn's kdtree implementation to very quickly get exact nearest neighbors. I would recommend that instead in this case.

**Comment by buhrmann** [comment]
Thanks for the quick response! And, sorry, I was actually thinking about the more general case of N dimensions and potentially large datasets (just forgot to include the n_components argument), and was wondering if there were any cached results from the fit() that would make extracting the nearest-neighbours quicker. At the moment I'm using spotify/annoy with the same metric as that used in UMAP to calculate those. I assume there shouldn't be any significant difference when compared to using umap_.nearest_neighbors then?

**Comment by lmcinnes** [comment]
No, no significant differences if that is the approach you want to use.

**Comment by sleighsoft** [comment]
@buhrmann Can this issue be closed?

**Comment by buhrmann** [comment]
Yes sure, sorry, should have closed it immediately...

**Comment by sleighsoft** [comment]
No worries, I am just trying to cleanup leftover issues :) Thanks for closing!
